### PR TITLE
feat: provide custom timeout for process termination

### PR DESCRIPTION
**Pull request by bazilmarkov**
- add termination timeout for graceful shutdown hook's processes after reaching the target
- send SIGKILL if the process is not responsive for SIGTERM

### CODE REVIEW

File: internal/handler/executor.go [hunk 1]
```diff
@@ -134,13 +138,49 @@ func (e *Executor) execHookCommand(w io.Writer) error {
...
+	if timeout > 0 {
+		// sets the same PGID for the child processes
+		setPGID(cmd)
+		terminationTimer := e.stopProcessWithTimeout(cmd, timeout)
+		err := cmd.Run()
+		// stop timer if command had finished before the timeout reached
+		terminationTimer.Stop()
+		return err
+	}
 	return cmd.Run()
```
Comment (dikderoy) [inline internal/handler/executor.go]: same as for killingTimer, and you can keep the exit point the same.


```suggestion
	if timeout > 0 {
		// sets the same PGID for the child processes
		setPGID(cmd)
		terminationTimer := e.stopProcessWithTimeout(cmd, timeout)
		defer func(t *time.Timer){
		    // stop timer if command had finished before the timeout reached
		    if !t.Stop() {
		        <-t.C
		    }
		}(terminationTimer)
	}
	return cmd.Run()
```

File: internal/middleware/dumper.go [hunk 2]
```diff
@@ -15,8 +15,8 @@ import (
 	"strings"
 )
 
-// responseDupper tees the response to a buffer and a response writer.
-type responseDupper struct {
+// responseDumper tees the response to a buffer and a response writer.
+type responseDumper struct {
```
Comment (dikderoy) [inline internal/middleware/dumper.go]: please revert this change. this struct is not responsible for the "dumping", it is responsible for "DUPlication" of the response. the dumper (after which the file is named) is the middleware. The difference in naming is intentional and shall not be changed.

also, it is out of scope for this MR.

File: internal/handler/executor.go [hunk 3]
```diff
@@ -105,6 +107,8 @@ func (e *Executor) execHookCommand(w io.Writer) error {
 	if err != nil {
 		return err
 	}
+	// retrieve timeout value
+	timeout := e.hook.Timeout * time.Second
```
Comment (dikderoy) [inline internal/handler/executor.go]: I think something wrong is going on here. with Timeout being of type `time.Duration` you should NOT multiply this by `time.Second`, as if I pass e.g. `10m` as duration (which is 10 * time.Minute) and then here you multiply this by `time.Second` you overflow the int64 type, which is a backing type for `time` units. Duration does not need to be manipulated at all, you should use it directly as is.

```suggestion
	timeout := e.hook.Timeout
```

File: internal/handler/executor.go [hunk 4]
```diff
@@ -134,13 +138,49 @@ func (e *Executor) execHookCommand(w io.Writer) error {
...
+			if err := sendKillSignal(e, cmd.Process.Pid, syscall.SIGKILL); err != nil {
+				e.logger.Error("Failed to send SIGKILL", "error", err)
+			}
+		})
+		// waiting for the process being exited
+		if processState, err := cmd.Process.Wait(); err != nil && processState != nil {
+			e.logger.Error("Error during process exiting", "error", err)
+		}
+		// stop the timer if process had terminated before timer reached
+		killingTimer.Stop()
```
Comment (dikderoy) [inline internal/handler/executor.go]: hoist to the timer creation point and refactor like this:

```suggestion
		killingTimer := time.AfterFunc(time.Second*time.Duration(10), func() {
			if err := sendKillSignal(e, cmd.Process.Pid, syscall.SIGKILL); err != nil {
				e.logger.Error("Failed to send SIGKILL", "error", err)
			}
		})
		defer func(t *time.Timer) {
		    // stop the timer if process had terminated before timer reached
  		    if !t.Stop() {
		        <-t.C
		    }
		}(killingTimer)

		// waiting for the process being exited
		if processState, err := cmd.Process.Wait(); err != nil && processState != nil {
			e.logger.Error("Error during process exiting", "error", err)
		}
```

**Review by dikderoy — CHANGES_REQUESTED** [review]
please review the comments. also, after consideration - regarding tests - you can actually test the new timeout configurations w/o the stream context. so you can just expand existing test suite, that should be possible. please give it a shot.

File: internal/handler/executor.go (hunk 3 above)
Comment (bazilmarkov) [reply internal/handler/executor.go:111]: I'm not sure if we can pass anything except int64 nanoseconds to default time.Duration realisation (unless we are operating Hook.Timeout as a string and use ParseDuration then)

Comment (bazilmarkov) [reply internal/handler/executor.go:111]: I'm not sure if we can pass anything except int64 nanoseconds to default `time.Duration` realisation through the `webhook.conf` (unless we are operating `Hook.Timeout` as a string and use `ParseDuration` then, which might be a more convenient way to represent timeout)

File: internal/handler/executor.go [hunk 5]
```diff
@@ -134,13 +138,60 @@ func (e *Executor) execHookCommand(w io.Writer) error {
...
+		setPGID(cmd)
+		terminationTimer := e.stopProcessWithTimeout(cmd, timeout)
+		defer func(t *time.Timer) {
+			// stop timer if command had finished before the timeout reached
+			t.Stop()
+			select {
+			case <-t.C:
+			default:
+			}
+		}(terminationTimer)
```
Comment (dikderoy) [inline internal/handler/executor.go]: my bad, seems like I overlooked this part of time.AfterFunc docs:

> The returned Timer's C field is not used and will be nil.

so it turns out that <-t.C will never yield as it is always nil for this object.
in that case it is enough to just call defer t.Stop() w/o any additional logic, which would've been necessary for raw time.Timer

```suggestion
        // stop the timer if process had terminated before timer reached
		defer terminationTimer.Stop()
```

File: internal/handler/executor.go [hunk 6]
```diff
@@ -134,13 +138,60 @@ func (e *Executor) execHookCommand(w io.Writer) error {
...
+			}
+		})
+		defer func(t *time.Timer) {
+			// stop the timer if process had terminated before timer reached
+			t.Stop()
+			select {
+			case <-t.C:
+			default:
+			}
+		}(killingTimer)
```
Comment (dikderoy) [inline internal/handler/executor.go]: same here.

> my bad, seems like I overlooked this part of time.AfterFunc docs:
>> The returned Timer's C field is not used and will be nil.
> so it turns out that <-t.C will never yield as it is always nil for this object.
> in that case it is enough to just call defer t.Stop() w/o any additional logic, which would've been necessary for raw time.Timer

```suggestion
                // stop the timer if process had terminated before timer reached
		defer killingTimer.Stop()
```

**Review by dikderoy — CHANGES_REQUESTED** [review]
besides those small changes, everything else looks good.

**Review by dikderoy — APPROVED** [review]

File: internal/handler/executor.go [hunk 7]
```diff
@@ -134,13 +138,48 @@ func (e *Executor) execHookCommand(w io.Writer) error {
...
+			e.logger.Warn("failed to send SIGTERM, trying SIGKILL instead", "error", err)
+		}
+		// attempting to kill the process with additional timeout on top
+		killingTimer := time.AfterFunc(time.Second*time.Duration(10), func() {
+			if err := sendKillSignal(e, cmd.Process.Pid, syscall.SIGKILL); err != nil {
+				e.logger.Error("failed to send SIGKILL", "error", err)
+			}
+		})
+                // stop the timer if process had terminated before timer reached
+		defer killingTimer.Stop()
```
Comment (dikderoy) [inline internal/handler/executor.go]:
```suggestion
        // stop the timer if process had terminated before timer reached
		defer killingTimer.Stop()
```
